### PR TITLE
feat: add code coverage

### LINKED ISSUE #50991: `istio-cni` code coverage
(This is used to request new product features, please visit <https://github.com/istio/istio/discussions> for questions on using Istio)

**Describe the feature request**
We don't have a good ongoing sense of what chunks have poor unit test coverage, and we don't have a good ongoing sense of where adding more cheap unit tests would bring the most value, and we generally want to encourage more, cheaper, easier to debug/maintain unit tests, and fewer, expensive, harder to debug/maintain integ/e2e tests.

We should generate code coverage stats (per branch/per func) as part of PR CI, and compare it with the current high-watermark in mainline, and warn/fail the CI check if coverage drops below the current high-watermark.

(usual caveat - code coverage is not _the only_ proof of good code coverage or useful tests, but it is the easiest to automate, and hard to casually cheat)

We could  do this for other components too , but `istio-cni` seems like a decent place to start as it is both relatively critical, relatively complex, and relatively self-contained.

**Unofficial requirements (subject to change)**

- Must work locally and in CI without extra configuration
- Must be something that can be invoked as a simple makefile target locally and in CI
- May not rely on external services being available
- Should dump coverage report as a CI artifact

When developing the inpod CNI internally we used overcover which is a very basic/simple thing that just parses go test coverage reports. It worked well enough and doesn't require any magic. Can always dump the coverage report as an artifact if people wanna see it, it doesn't need to be fancy.

overcover --coverprofile cover.out ./... --threshold $(COVERAGE_THRESH_PCT)

tl;dr: I really don't want to use anything that can't work locally AND as a simple makefile target.

EDIT: After 5/15 WG, following are the requirements:

- Start with one component at a time (`istio-cni` in this case)
- Simple, language-specific tooling only, must run locally and in CI the same way (makefile target, no out of band setup)
- CI should publish coverage report artifacts.
- An overall number for _all_ the packages in the component should be generated (`go test` needs help to do this IIRC or it will only report per-go-pkg coverages, which is not as useful - something like [`overcover`](https://github.com/klmitch/overcover) can help with that)
- Stretch goal -> figure out how to get integration tests  to count towards coverage. Istio currently has an (IMO excessively) topheavy inverted test pyramid, so being able to include integ tests in coverage would be nice too).

### PR DESCRIPTION
**Pull request by chansuke**
Closes #50991

### CODE REVIEW

File: .github/workflows/coverage.yaml [hunk 1]
```diff
@@ -0,0 +1,21 @@
+jobs:
+  coverage:
+    runs-on: ubuntu-latest
```
[the remark this answers is not in the document]
Comment (howardjohn) [reply .github/workflows/coverage.yaml:3]: Dumping a file to $ARTIFACTS will upload it in prow

**Review by howardjohn — CHANGES_REQUESTED** [review]

**Review by my-git9 — COMMENTED** [review]
/test release-notes

**Review by bleggett — APPROVED** [review]
This LGTM (it at least provides a manual/local coverage threshold check, and a coverage report for prow).

Thanks!

File: prow/config/coverage.yaml
Comment (howardjohn) [inline prow/config/coverage.yaml:1]: Do we need this?

File: prow/coverage.sh [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+      echo "Overcover is not installed. Installing..."
+      go install github.com/klmitch/overcover@latest
+    fi
+    # Run Overcover to validate coverage threshold
+    overcover --coverprofile="${REPORT_COVERAGE}" ./cni/... --threshold="${COVERAGE_THRESH_PCT}"
+    echo "Coverage report generated: ${REPORT_COVERAGE}"
+    ;;
+  report)
+    # Upload the coverage report to GCS
+    gsutil cp "${REPORT_COVERAGE}" "gs://${GCS_BUCKET:-istio-prow}/coverage/${GIT_SHA}.out"
```
Comment (howardjohn) [inline prow/coverage.sh:46]: I don't know that we need to do this, just dump a file into $ARTIFACTS and prow will automatically include it.

File: prow/coverage.sh [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+WD=$(cd "$WD"; pwd)
+ROOT=$(dirname "$WD")
+
+# shellcheck source=prow/lib.sh
+source "${ROOT}/prow/lib.sh"
+setup_and_export_git_sha
+
+# Define paths for reports
+REPORT_COVERAGE="${REPORT_COVERAGE:-${ARTIFACTS}/coverage.out}"
+COVERAGE_THRESH_PCT="${COVERAGE_THRESH_PCT:-50}" # Default coverage threshold
```
Comment (howardjohn) [inline prow/coverage.sh:29]: I am not sure what the point of a threshold is given we are not going to enforce it

Comment (howardjohn) [reply prow/coverage.sh:29]: To clarify -- there are different opinions from different people on the usefulness of reporting coverage in general, but adding coverage gates has been explicitly rejected.

File: prow/coverage.sh (hunk 2 above)
Comment (chansuke) [reply prow/coverage.sh:46]: ok

File: prow/config/coverage.yaml
Comment (chansuke) [reply prow/config/coverage.yaml:1]: Oh, we don't need this

File: prow/coverage.sh [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+setup_and_export_git_sha
+
+export ARTIFACTS="${ARTIFACTS:-$(mktemp -d)}"
+
+# Define paths for reports
+REPORT_COVERAGE="${REPORT_COVERAGE:-${ARTIFACTS}/coverage.out}"
+
+mkdir -p "$(dirname "${REPORT_COVERAGE}")"
+# Run the Go coverage test
+go test ./... -coverprofile="${REPORT_COVERAGE}" || true
```
Comment (howardjohn) [inline prow/coverage.sh:34]: Do we really want to ignore failures?

File: prow/coverage.sh [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+if ! command -v overcover &>/dev/null; then
+  echo "Overcover is not installed. Installing..."
+  go install github.com/klmitch/overcover@latest
+fi
+# Run Overcover to validate coverage threshold
+overcover --coverprofile="${REPORT_COVERAGE}" ./cni/...
```
Comment (howardjohn) [inline prow/coverage.sh:40]: Why do we need ./... and ./cni/...

**Comment by chansuke** [comment]
@howardjohn 
friendly ping

**Review by howardjohn — APPROVED** [review]
